src/rascal/structure_managers/adaptor_increase_maxorder.hh

```diff
@@ -459,28 +430,84 @@ namespace rascal {
+        AddOrderLoop<2, traits::NeighbourListType, IsCompactCluster, true>;
+    static constexpr bool has_center_pair_order_two{traits::HasCenterPair and
+                                                    traits::MaxOrder - 1 == 2};
+    // size_t turns out to give 0 if false and 1 if true. When adding Order==3
+    // to the manager the input manager could have center pairs and
+    // cluster_start avoid iterating over those
+    static constexpr size_t cluster_start{
+        static_cast<size_t>(has_center_pair_order_two)};
 
     internal::for_each(this->cluster_indices_container,
                        internal::ResizePropertyToZero());
 
+    auto & atom_cluster_indices{std::get<0>(this->cluster_indices_container)};
+
     this->nb_neigh.clear();
     this->offsets.clear();
     this->neighbours_atom_tag.clear();
 
-    // #BUG8486@(markus) I now append the ghost atoms to the cluster index
-    // container
     for (auto atom : this->manager) {
-      //  Order 1, but variable Order is at 0, atoms, index 0
-      using AddOrderLoop = AddOrderLoop<atom.order(), traits::NeighbourListType,
-                                        IsCompactCluster,
-                                        atom.order() == (traits::MaxOrder - 1)>;
-
-      auto & atom_cluster_indices{std::get<0>(this->cluster_indices_container)};
-
-      auto indices = atom.get_cluster_indices();
+      // copy the underlying cluster indices of this->manager (not changed)
+      // do order == 1
+      auto indices{atom.get_cluster_indices()};
       atom_cluster_indices.push_back(indices);
-
-      AddOrderLoop::loop(atom, atom, 0, *this);
+      // do the rest
+      AddOrderLoop::loop(atom, *this);
+
+      this->add_entry_number_of_neighbours();
+      std::array<int, traits::MaxOrder> new_tag_list{};
+      // loop over the highest order available in this->manager
+      for (auto cluster :
+           atom.template get_clusters_of_order<traits::MaxOrder - 1>(
+               cluster_start)) {
+        auto && tag_list{cluster.get_atom_tag_list()};
+        // copy the tags from the previous order skiping the center atom tag
+        for (size_t ii{0}; ii < traits::MaxOrder - 1; ++ii) {
+          new_tag_list[ii] = tag_list[ii];
+        }
+        for (auto pair : atom.template get_clusters_of_order<2>()) {
+          // copy to the last element the atom tag of the new order
+          new_tag_list.back() = pair.get_atom_tag();
+          if (traits::NeighbourListType ==
+              AdaptorTraits::NeighbourListType::half) {
+            if (IsCompactCluster) {
+              throw std::runtime_error("Not implemented yet.");
+            } else {
+              // only add strictly lexicographicaly ordered tags, e.g.
+              // (1,2,2) or (1,5,3) are not valid and (1,3,5) is valid
+              if (new_tag_list[traits::MaxOrder - 1] >
+                  new_tag_list[traits::MaxOrder - 2]) {
+                std::array<int, traits::MaxOrder - 1> tags{};
```

Comment:
This seems to suggest that it is only valid for up to triplets?
Or are you expecting the lower order tags to be already sorted?

I still think there is maybe a better solution, I just don't know which one.
Just stuff I keep thinking about:
Suppose I make a neighbourlist (minimal, so ordered), then I reorder them according to distances (like in your example), then I build the triplets. From there I change the reordering of the triplets to some sorting criterion (minimal radius) and then try to build the quadruplets from there. This would be caught here, right?

Reply:
>This seems to suggest that it is only valid for up to triplets? 

since increase max order has been used to make the triplet, at this point the tag of the triplet will be sorted and you only need to check for the last tag with the quadruplet tag.

> This would be caught here, right?

it will be caught here because the reordering happens between clusters and not the tags within the clusters.

Reply:
I see.